### ISSUE TITLE
new lint: Unused trait bound(s)

### ISSUE THREAD

**Issue by taufik-rama** [rust-lang/rust-clippy#9316]
### What it does

For any trait bounds that's not used on the function body, create an "unused" warning

### Lint Name

unused_trait_bound

### Category

pedantic

### Advantage

- More "tight" code, we can be sure that each bounds will be used
  - and maybe in other words: more cleaner code since unused bound can be safely deleted

### Drawbacks

- Might not be suitable for library APIs, since removing bounds (even if it's unused) might be a breaking change
  - It might need some thought on how it would interact with [`negative_impls`](https://doc.rust-lang.org/unstable-book/language-features/negative-impls.html)

### Example

```rust
trait Database {
    fn database(&self);
}

trait Redis {
    fn redis(&self);
}

fn get_user<C>(conn: C)
where
    C: Database + Redis, // this `Redis` bound should be warned
{
    // we just so happen to forgot to check for Redis cache
    conn.database()
}
```

Could be written as:

```rust
trait Database {
    fn database(&self);
}

trait Redis {
    fn redis(&self);
}

fn get_user<C>(conn: C)
where
    C: Database, // either remove the bound, or
{
    // use the bound
    // conn.redis()
    conn.database()
}
```

**Comment by yvt** [comment]
> * Might not be suitable for library APIs, since removing bounds (even if it's unused) might be a breaking change

It's the other way around - adding bounds is the breaking change. Removing bounds is a non-breaking change but might impose design constraints in future semver-compatible versions of the libraries.

**Comment by taufik-rama** [comment]
Yeah, for more common trait constraints I would guess removing it would means more code should be accepted. I was thinking of stuff like `!Sized` and the like, though it might be too pessimistic in my part haha

**Comment by leonardo-m** [comment]
See also:
https://github.com/rust-lang/rust-clippy/issues/3216

**Comment by taufik-rama** [comment]
Ah, I didn't find that issue when I searched about this, but I did expect this to already be suggested before

I don't know how to proceed, should I close this issue?

**Comment by xFrednet** [comment]
If it's a duplicate, then it would be nice to close one of them. This one seems more details and has an example. Therefore, I'll close the older one. @leonardo-m you can ping me or reopen it if you believe that there is a difference in the suggested lints :upside_down_face:

**Comment by leonardo-m** [comment]
Closing the old one is right.
That one was from 2018, I felt it was a basic warning to have, so it was surprising to see it open for so much time. I remove useless bounds all the time, and I remember sevaral rustc pull requests to remove post-facto trait bounds in the stdlib. So I still think it's going to be a quite useful lint (perhaps worth having in rusts itself).

**Comment by jokeyrhyme** [comment]
See also #1463
